_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,440 @@
+
+### Initializing Mocks with Mockito Annotations
+
+If the system under test has several dependencies that must be mocked, it gets cumbersome to create all these mocks with
+the variant shown above. 
+
+To initialize them all at once, we can annotate them with `@Mock`:
+
+```java
+class CityServiceImplTestMockitoAnnotationStyle {
+
+  // System under Test (SuT)
+  private CityService cityService;
+
+  // Mocks
+  @Mock
+  private CityRepository cityRepository;
+
+  @BeforeEach
+  void setUp() {
+    MockitoAnnotations.openMocks(this);
+    cityService = new CityServiceImpl(cityRepository);
+  }
+
+}
+```
+
+Applying this variant, we don't have to deal with boilerplate code and are able to keep our unit test neat and concise.
+`MockitoAnnotations.openMocks(this)` initializes the fields annotated with `@Mock` for the given test class, which is in
+our case the class itself.
+
+### Delegating the Initialization to an Extension
+
+As an alternative to the Mockito annotation style we can make use of JUnit Jupiter's `@ExtendWith` and extend JUnit
+Jupiter's context with `MockitoExtension.class`:
+
+```java
+@ExtendWith(MockitoExtension.class)
+class CityServiceImplTestMockitoJUnitExtensionStyle {
+
+  // System under Test (SuT)
+  private CityService cityService;
+
+  // Mocks
+  @Mock
+  private CityRepository cityRepository;
+
+  @BeforeEach
+  void setUp() {
+    cityService = new CityServiceImpl(cityRepository);
+  }
+
+}
+```
+The extension assumes the initialization for annotated fields, so we must not do it ourselves. This makes our setup
+even neater and conciser!
+
+### Injecting Mocks with Spring
+
+If we have a more complex test fixture, and we want to inject the mock into Spring's `ApplicationContext` we can make
+use of `@MockBean`:
+
+```java
+@ExtendWith(SpringExtension.class)
+class CityServiceImplTestMockitoSpringStyle {
+
+  // System under Test (SuT)
+  private CityService cityService;
+
+  // Mocks
+  @MockBean
+  private CityRepository cityRepository;
+
+  @BeforeEach
+  void setUp() {
+    cityService = new CityServiceImpl(cityRepository);
+  }
+
+}
+```
+
+But caution: `@MockBean` is not an annotation from Mockito but from Spring!
+In the startup process Spring places the mock in the context, so that we don't need to do it ourselves.
+Wherever a bean claims to have its dependency satisfied, Spring injects the mock instead of the real object.
+This become handy if we want to have the same mock in different places.
+
+See [Mocking with Mockito and Spring Boot](https://reflectoring.io/spring-boot-mock/#mocking-with-mockito-and-spring-boot)
+for a deep dive how to mock Beans in Spring Boot.
+
+## Mockito Best Practices
+
+Knowing how to create the mocks, let's have a look at what we should avoid. Otherwise, our test cases are confusing
+what decreases our code's maintainability. It will save us much time debugging and doesn't let our team members guess
+what the intent of the test case is.
+
+### Avoid Concatenation in `setUp()`
```

Comment:
I know what you mean to say in this section, but it's not very clear. You're starting the section like everybody knows what you're talking about. Please re-write this section to make it clearer. Also, what do I do instead to make it better? What's the best practice?